File: experiments/adrianco/experiment-13/runs/language=go_model=opus-4.8-fast_prompt=ATDD/rep3/format.go
// Context: Presentation layer for the Brazilian Soccer MCP server. These
// functions turn query results from the store into the human-readable, domain
// language text that each MCP tool returns to the calling LLM (match lists with
// scores, team records, head-to-head tallies, league standings, aggregate
// statistics and player listings). Keeping formatting here keeps the store
// purely about data and the tools purely about argument handling.
package main

import (
	"fmt"
	"strings"
)

const maxListed = 50

// plural renders "1 win" / "2 wins" (and "1 match" / "2 matches").
func plural(n int, noun string) string {
	if n == 1 {
		return fmt.Sprintf("%d %s", n, noun)
	}
	if noun == "match" {
		return fmt.Sprintf("%d matches", n)
	}
	return fmt.Sprintf("%d %ss", n, noun)
}

// matchLine renders a single match result line.
func matchLine(m Match) string {
	score := "?-?"
	if m.HasScore {
		score = fmt.Sprintf("%d-%d", m.HomeGoals, m.AwayGoals)
	}
	extra := m.Competition
	switch {
	case m.Stage != "":
		extra += " - " + m.Stage
	case m.Round != "":
		extra += " Round " + m.Round
	}
	return fmt.Sprintf("- %s: %s %s %s (%s)", m.dateString(), m.HomeTeam, score, m.AwayTeam, extra)
}

func listMatches(matches []Match) string {
	var b strings.Builder
	limit := len(matches)
	if limit > maxListed {
		limit = maxListed
	}
	for i := 0; i < limit; i++ {
		b.WriteString(matchLine(matches[i]))
		b.WriteByte('\n')
	}
	if len(matches) > limit {
		fmt.Fprintf(&b, "... and %d more\n", len(matches)-limit)
	}
	return b.String()
}

// formatFindMatches renders the result of a match search. When both a team and
// an opponent were specified it appends a head-to-head summary.
func formatFindMatches(label string, matches []Match, h2h *HeadToHead, aName, bName string) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Found %s for %s:\n", plural(len(matches), "match"), label)
	if len(matches) == 0 {
		b.WriteString("(no matches found in the dataset)\n")
		return b.String()
	}
	b.WriteString(listMatches(matches))
	if h2h != nil {
		fmt.Fprintf(&b, "\nHead-to-head (these teams): %s %s, %s %s, %s\n",
			aName, plural(h2h.AWins, "win"),
			bName, plural(h2h.BWins, "win"),
			plural(h2h.Draws, "draw"))
	}
	return b.String()
}

func formatTeamRecord(team string, season int, hasSeason bool, competition, venue string, r TeamRecord) string {
	scope := venueLabel(venue) + " record"
	var quals []string
	if hasSeason {
		quals = append(quals, itoa(season))
	}
	if competition != "" {
		quals = append(quals, competition)
	}
	header := fmt.Sprintf("%s %s", team, scope)
	if len(quals) > 0 {
		header += " (" + strings.Join(quals, ", ") + ")"
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%s:\n", header)
	fmt.Fprintf(&b, "- Matches: %d\n", r.Matches)
	fmt.Fprintf(&b, "- Wins: %d, Draws: %d, Losses: %d\n", r.Wins, r.Draws, r.Losses)
	fmt.Fprintf(&b, "- Goals For: %d, Goals Against: %d\n", r.GoalsFor, r.GoalsAgainst)
	fmt.Fprintf(&b, "- Win rate: %.1f%%\n", r.WinRate())
	return b.String()
}

func venueLabel(venue string) string {
	switch venue {
	case "home":
		return "home"
	case "away":
		return "away"
	default:
		return "overall"
	}
}

func formatHeadToHead(aName, bName string, h HeadToHead) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Head-to-head: %s vs %s\n", aName, bName)
	fmt.Fprintf(&b, "Total matches: %d\n", len(h.Matches))
	fmt.Fprintf(&b, "- %s wins: %d\n", aName, h.AWins)
	fmt.Fprintf(&b, "- %s wins: %d\n", bName, h.BWins)
	fmt.Fprintf(&b, "- Draws: %d\n", h.Draws)
	fmt.Fprintf(&b, "Goals: %s %d, %s %d\n", aName, h.AGoals, bName, h.BGoals)
	if len(h.Matches) > 0 {
		b.WriteString("\nMatches:\n")
		b.WriteString(listMatches(h.Matches))
	}
	return b.String()
}

func formatStandings(competition string, season int, table []Standing) string {
	var b strings.Builder
	fmt.Fprintf(&b, "%s %d Final Standings (calculated from matches):\n", competition, season)
	if len(table) == 0 {
		b.WriteString("(no matches found for this competition and season)\n")
		return b.String()
	}
	for i, st := range table {
		fmt.Fprintf(&b, "%d. %s - %d pts (%dW, %dD, %dL) GF:%d GA:%d GD:%+d\n",
			i+1, st.Team, st.Points, st.Wins, st.Draws, st.Losses,
			st.GoalsFor, st.GoalsAgainst, st.GoalDiff())
	}
	return b.String()
}

func formatStatistics(scope string, stat Statistics) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Statistics (%s):\n", scope)
	fmt.Fprintf(&b, "- Matches: %d\n", stat.Matches)
	fmt.Fprintf(&b, "- Total goals: %d\n", stat.TotalGoals)
	fmt.Fprintf(&b, "- Average goals per match: %.2f\n", stat.AvgGoals())
	fmt.Fprintf(&b, "- Home win rate: %.1f%%\n", stat.HomeWinRate())
	fmt.Fprintf(&b, "- Draw rate: %.1f%%\n", stat.DrawRate())
	fmt.Fprintf(&b, "- Away win rate: %.1f%%\n", stat.AwayWinRate())
	if len(stat.BiggestWins) > 0 {
		b.WriteString("\nBiggest victories:\n")
		for i, m := range stat.BiggestWins {
			margin := abs(m.HomeGoals - m.AwayGoals)
			fmt.Fprintf(&b, "%d. %s: %s %d-%d %s (margin %d)\n",
				i+1, m.dateString(), m.HomeTeam, m.HomeGoals, m.AwayGoals, m.AwayTeam, margin)
		}
	}
	return b.String()
}

func formatPlayers(players []Player, limit int) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Found %s:\n", plural(len(players), "player"))
	if len(players) == 0 {
		b.WriteString("(no players found in the dataset)\n")
		return b.String()
	}
	shown := len(players)
	if limit > 0 && limit < shown {
		shown = limit
	}
	for i := 0; i < shown; i++ {
		p := players[i]
		fmt.Fprintf(&b, "%d. %s - Overall: %d, Position: %s, Club: %s, Nationality: %s\n",
			i+1, p.Name, p.Overall, dashIfEmpty(p.Position), dashIfEmpty(p.Club), dashIfEmpty(p.Nationality))
	}
	if len(players) > shown {
		fmt.Fprintf(&b, "... and %d more\n", len(players)-shown)
	}
	return b.String()
}

func dashIfEmpty(s string) string {
	if strings.TrimSpace(s) == "" {
		return "-"
	}
	return s
}

File: experiments/adrianco/experiment-7/brazil/runs/language=clojure_model=claude-opus-4-8-fast/rep3/src/brazilian_soccer/format.clj
;; =============================================================================
;; brazilian-soccer.format
;; -----------------------------------------------------------------------------
;; CONTEXT
;;   Turns the plain-data results from brazilian-soccer.queries into the
;;   human-readable text blocks shown in TASK.md's "Example answer format"
;;   sections. The MCP server returns these strings as tool-call text content so
;;   the connected LLM (and a human reading the transcript) sees tidy answers.
;;
;;   Kept separate from queries.clj so the analytics stay pure/testable and the
;;   presentation can evolve independently.
;; =============================================================================
(ns brazilian-soccer.format
  (:require [clojure.string :as str]
            [brazilian-soccer.queries :as q]))

(defn- pct [x] (format "%.1f%%" (* 100.0 (double x))))

(defn- score [m]
  (str (or (:home-goals m) "?") "-" (or (:away-goals m) "?")))

(defn- comp-tag [m]
  (let [bits (remove nil? [(:competition m)
                           (when (:round m) (str "Round " (:round m)))
                           (:stage m)
                           (:season m)])]
    (str "(" (str/join ", " bits) ")")))

(defn match-line [m]
  (str (or (:date m) "????-??-??") ": "
       (:home m) " " (score m) " " (:away m) " " (comp-tag m)))

;; ---------------------------------------------------------------------------

(defn format-matches [matches {:keys [title]}]
  (if (empty? matches)
    "No matches found for that query."
    (str (when title (str title "\n"))
         (str/join "\n" (map #(str "- " (match-line %)) matches))
         "\n\nTotal: " (count matches) " match(es).")))

(defn format-head-to-head [h]
  (if (zero? (:total h))
    (str "No matches found between " (:team-a h) " and " (:team-b h) ".")
    (str (:team-a h) " vs " (:team-b h) " — head-to-head (provided data):\n"
         "- Matches: " (:total h) "\n"
         "- " (:team-a h) " wins: " (:a-wins h) "\n"
         "- " (:team-b h) " wins: " (:b-wins h) "\n"
         "- Draws: " (:draws h) "\n"
         "- Goals: " (:team-a h) " " (:a-goals h) " — " (:b-goals h) " " (:team-b h) "\n\n"
         "Recent meetings:\n"
         (str/join "\n" (map #(str "- " (match-line %)) (take 10 (:matches h)))))))

(defn format-team-stats [s]
  (if (nil? s)
    "No matches found for that team / filters."
    (let [scope (->> [(when (:competition s) (:competition s))
                      (when (:season s) (str "season " (:season s)))]
                     (remove nil?) (str/join ", "))
          wr (if (pos? (:matches s)) (/ (:wins s) (double (:matches s))) 0.0)]
      (str (:team s) " record" (when (seq scope) (str " (" scope ")")) ":\n"
           "- Matches: " (:matches s) "\n"
           "- Wins: " (:wins s) ", Draws: " (:draws s) ", Losses: " (:losses s) "\n"
           "- Goals For: " (:goals-for s) ", Goals Against: " (:goals-against s)
           " (diff " (let [d (- (:goals-for s) (:goals-against s))] (if (pos? d) (str "+" d) d)) ")\n"
           "- Win rate: " (pct wr)))))

(defn format-players [players {:keys [title]}]
  (if (empty? players)
    "No players found for that query."
    (str (when title (str title "\n"))
         (str/join "\n"
                   (map-indexed
                    (fn [i p]
                      (format "%d. %s — Overall: %s, Position: %s, Club: %s%s"
                              (inc i) (:name p) (or (:overall p) "?")
                              (or (:position p) "?") (or (:club p) "Free agent")
                              (if (:nationality p) (str " [" (:nationality p) "]") "")))
                    players))
         "\n\nShowing " (count players) " player(s).")))

(defn format-standings [rows {:keys [competition season]}]
  (if (empty? rows)
    (str "No standings available for " competition " " season ".")
    (str competition " " season " — table (calculated from matches):\n"
         (str/join "\n"
                   (map (fn [r]
                          (format "%2d. %-22s %3d pts  (%2dW %2dD %2dL, GF %d GA %d, GD %+d)"
                                  (:rank r) (:team r) (:points r)
                                  (:wins r) (:draws r) (:losses r)
                                  (:gf r) (:ga r) (:gd r)))
                        rows)))))

(defn format-league-stats [s {:keys [competition season]}]
  (let [scope (->> [competition (when season (str season))]
                   (remove nil?) (str/join " "))]
    (str "Statistics" (when (seq scope) (str " for " scope)) ":\n"
         "- Matches: " (:matches s) " (with scores: " (:scored-matches s) ")\n"
         "- Total goals: " (:total-goals s) "\n"
         "- Average goals per match: " (format "%.2f" (:avg-goals s)) "\n"
         "- Home win rate: " (pct (:home-win-rate s)) "\n"
         "- Away win rate: " (pct (:away-win-rate s)) "\n"
         "- Draw rate: " (pct (:draw-rate s)))))

(defn format-biggest-wins [matches {:keys [title]}]
  (if (empty? matches)
    "No matches found for that query."
    (str (or title "Biggest victories (provided data):") "\n"
         (str/join "\n"
                   (map-indexed
                    (fn [i m] (str (inc i) ". " (match-line m)
                                   "  [margin " (:margin m) "]"))
                    matches)))))

(defn format-competitions [comps]
  (str "Competitions in the knowledge graph:\n"
       (str/join "\n"
                 (map (fn [c]
                        (let [ss (:seasons c)]
                          (format "- %s: %d matches, %d teams, seasons %s–%s"
                                  (:competition c) (:matches c) (:teams c)
                                  (str (first ss)) (str (last ss)))))
                      comps))))
